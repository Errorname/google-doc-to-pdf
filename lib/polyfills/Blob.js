"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const node_fetch_1 = require("node-fetch");
const realBlob = typeof window !== 'undefined' ? window.Blob : node_fetch_1.Blob;
exports.default = realBlob;
