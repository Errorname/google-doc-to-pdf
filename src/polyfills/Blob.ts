import { Blob as BlobNode } from 'node-fetch'

const realBlob = typeof window !== 'undefined' ? window.Blob : BlobNode

export default realBlob
