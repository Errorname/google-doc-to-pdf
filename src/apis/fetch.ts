import crossFetch from 'cross-fetch'
import Blob from '../polyfills/Blob'

export const fetch = (token: Function) => (
  url: string,
  body: any = null,
  options: { method?: string; headers?: any; raw?: boolean } = {}
): Promise<Response> => {
  const { method, headers, raw, ...rest } = options

  if (body && !(body instanceof Blob)) {
    body = JSON.stringify(body)
  }

  return crossFetch(url, {
    method: method || body ? 'POST' : 'GET',
    body,
    headers: {
      authorization: `Bearer ${token()}`,
      'Content-Type': 'application/json',
      ...headers
    },
    ...rest
  }).then(r => (raw ? r : r.json()))
}

export const multipart = (
  parts: {
    data: string
    [propName: string]: string
  }[],
  boundary: string
): Blob => {
  const body = []

  parts.map(part => {
    const { data, ...headers } = part
    body.push(
      [
        `--${boundary}`,
        ...Object.entries(headers).map(([key, value]) => `${key}: ${value}`),
        '\n'
      ].join('\n')
    )
    body.push(data)
    body.push('\n\n')
  })

  body.push(`--${boundary}--`)

  return new Blob(body)
}
