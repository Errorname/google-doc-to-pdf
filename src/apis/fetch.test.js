import { fetch, multipart } from './fetch'
import crossFetch from 'cross-fetch'
import Blob from '../polyfills/Blob'

jest.mock('cross-fetch', () => jest.fn())
jest.mock('../polyfills/Blob')

describe('fetch', () => {
  const token = async () => 'my-token'

  beforeEach(() => {
    crossFetch.mockReset()
    crossFetch.mockResolvedValue({ status: 200, json: () => Promise.resolve({ some: 'data' }) })
  })

  test('mimimum', async () => {
    const result = await fetch(token)('the-url')

    expect(crossFetch).toHaveBeenCalled()
    expect(crossFetch.mock.calls).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })

  test('with body', async () => {
    await fetch(token)('the-url', { test: 'test' })

    expect(crossFetch).toHaveBeenCalled()
    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('custom method', async () => {
    await fetch(token)('the-url', { test: 'test' }, { method: 'DELETE' })

    expect(crossFetch).toHaveBeenCalled()
    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('with headers', async () => {
    await fetch(token)('the-url', { test: 'test' }, { headers: { 'X-TEST': '1234' } })

    expect(crossFetch).toHaveBeenCalled()
    expect(crossFetch.mock.calls).toMatchSnapshot()
  })

  test('raw response', async () => {
    const result = await fetch(token)('the-url', { test: 'test' }, { raw: true })

    expect(crossFetch).toHaveBeenCalled()
    expect(crossFetch.mock.calls).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })

  test('http error', async () => {
    crossFetch.mockReset()
    crossFetch.mockResolvedValue({ status: 500, json: () => Promise.resolve({ some: 'data' }) })

    expect(fetch(token)('the-url')).rejects.toThrowErrorMatchingSnapshot()
  })
})

describe('multipart', () => {
  beforeAll(() => {
    Blob.mockImplementation(args => ({ args: args.join('') }))
  })

  test('minimum', () => {
    const blob = multipart([{ data: '1234' }], 'boundary1')

    expect(blob.args).toMatchSnapshot()
  })

  test('multiple parts', () => {
    const blob = multipart([{ data: '1234' }, { data: 'abcd' }], 'boundary1')

    expect(blob.args).toMatchSnapshot()
  })

  test('with headers', () => {
    const blob = multipart([{ data: '1234', Authorization: 'Bearer the-token' }], 'boundary1')

    expect(blob.args).toMatchSnapshot()
  })
})
