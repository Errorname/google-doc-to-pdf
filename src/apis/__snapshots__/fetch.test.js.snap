// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch custom method 1`] = `
Array [
  Array [
    "the-url",
    Object {
      "body": "{\\"test\\":\\"test\\"}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "DELETE",
    },
  ],
]
`;

exports[`fetch mimimum 1`] = `
Array [
  Array [
    "the-url",
    Object {
      "body": null,
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`fetch mimimum 2`] = `
Object {
  "status": 200,
}
`;

exports[`fetch raw response 1`] = `
Array [
  Array [
    "the-url",
    Object {
      "body": "{\\"test\\":\\"test\\"}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetch raw response 2`] = `
Object {
  "json": [Function],
}
`;

exports[`fetch with body 1`] = `
Array [
  Array [
    "the-url",
    Object {
      "body": "{\\"test\\":\\"test\\"}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetch with headers 1`] = `
Array [
  Array [
    "the-url",
    Object {
      "body": "{\\"test\\":\\"test\\"}",
      "headers": Object {
        "Content-Type": "application/json",
        "X-TEST": "1234",
        "authorization": "Bearer my-token",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`multipart minimum 1`] = `
"--boundary1

1234

--boundary1--"
`;

exports[`multipart multiple parts 1`] = `
"--boundary1

1234

--boundary1

abcd

--boundary1--"
`;

exports[`multipart with headers 1`] = `
"--boundary1
Authorization: Bearer the-token

1234

--boundary1--"
`;
